packages/deployments/contracts/contracts/diamond/facets/StableSwapFacet.sol

```diff
@@ -0,0 +1,457 @@
+ *
+ * Each time a swap between the pooled tokens happens, a set fee incurs which effectively gets
+ * distributed to the LPs.
+ *
+ * In case of emergencies, admin can pause additional deposits, swaps, or single-asset withdraws - which
+ * stops the ratio of the tokens in the pool from changing.
+ * Users can always withdraw their tokens via multi-asset withdraws.
+ *
+ * @dev Most of the logic is stored as a library `SwapUtils` for the sake of reducing contract's
+ * deployment size.
+ */
+
+contract StableSwapFacet is Modifiers {
+  using SwapUtils for SwapUtils.Swap;
+  using AmplificationUtils for SwapUtils.Swap;
+
+  // ========== Custom Errors ===========
+  error StableSwapFacet__deadlineCheck_deadlineNotMet();
+  error StableSwapFacet__poolCheck_notConfigured();
+  error StableSwapFacet__poolCheck_alreadyInitialized();
+  error StableSwapFacet__initializeSwap_invalidPooledTokens();
+  error StableSwapFacet__initializeSwap_decimalsMismatch();
+  error StableSwapFacet__initializeSwap_duplicateTokens();
+  error StableSwapFacet__initializeSwap_zeroTokenAddress();
+  error StableSwapFacet__initializeSwap_tokenDecimalsExceedMax();
+  error StableSwapFacet__initializeSwap_aExceedMax();
+  error StableSwapFacet__initializeSwap_feeExceedMax();
+  error StableSwapFacet__initializeSwap_adminFeeExceedMax();
+  error StableSwapFacet__initializeSwap_failedInitLpTokenClone();
+  error StableSwapFacet__getToken_outOfRange();
+  error StableSwapFacet__getTokenIndex_notExist();
+  error StableSwapFacet__getTokenBalance_indexOutOfRange();
+
+  // ============ Properties ============
+
+  // ============ Modifiers ============
+
+  /**
+   * @notice Modifier to check deadline against current timestamp
+   * @param deadline latest timestamp to accept this transaction
+   */
+  modifier deadlineCheck(uint256 deadline) {
+    if (block.timestamp > deadline) revert StableSwapFacet__deadlineCheck_deadlineNotMet();
+    _;
+  }
+
+  /**
+   * @notice Initializes this Swap contract with the given parameters.
+   * This will also clone a LPToken contract that represents users'
+   * LP positions. The owner of LPToken will be this contract - which means
+   * only this contract is allowed to mint/burn tokens.
+   *
+   * @param _canonicalId the canonical token id
+   * @param _pooledTokens an array of ERC20s this pool will accept
+   * @param decimals the decimals to use for each pooled token,
+   * eg 8 for WBTC. Cannot be larger than POOL_PRECISION_DECIMALS
+   * @param lpTokenName the long-form name of the token to be deployed
+   * @param lpTokenSymbol the short symbol for the token to be deployed
+   * @param _a the amplification coefficient * n * (n - 1). See the
+   * StableSwap paper for details
+   * @param _fee default swap fee to be initialized with
+   * @param _adminFee default adminFee to be initialized with
+   * @param lpTokenTargetAddress the address of an existing LPToken contract to use as a target
+   */
+  function _initializeSwap(
```

Comment:
I think this logic needs to be moved to `DiamondInit.init` function

Reply:
I'm not sure exactly how it is supposed to work, but my guess is that there will be multiple initializations, one per each stableswap instance created? So probably this `_initializeSwap` should be called from an external method that allows to create a new/initialize a new stableSwap?

Reply:
> 

yes, exactly, initializeSwap is to init a swap for a canonical token.
I will fix